### PR TITLE
Fix committee validation in Ratify

### LINKED ISSUE #3889: Current validation of UpdateCommitte action is incorrect
On this line we use current CommitteMembers for validation:

https://github.com/input-output-hk/cardano-ledger/blob/afedb7d519761ccdd9c013444aa4b3e0bf0e68ef/eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs#L312

Which is incorrect. We must be using the new committee members that are being proposed.

Here is the corresponding spec part of the code:

https://github.com/input-output-hk/formal-ledger-specifications/blob/eafd83a7931cd6af07212253a183115e740f6c1e/src/Ledger/GovernanceActions.lagda#L308-L311

### PR DESCRIPTION
**Pull request by teodanciu**
# Description

Closes https://github.com/input-output-hk/cardano-ledger/issues/3889

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs [hunk 1]
```diff
@@ -395,14 +394,17 @@ updatePrevGovActionIdsChildren gas pgac =
 
 validCommitteeTerm ::
   ConwayEraPParams era =>
-  StrictMaybe (Committee era) ->
   PParams era ->
   EpochNo ->
+  GovAction era ->
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: Not sure whether to mention the change of the signature in CHANGELOG, since it doesn't appear there to begin with (it's exported for testing). I suppose people could still use it, though.
If that's the case, maybe it's not worth changing the order of these parameters? I only did it so I can use the lambda case, but can't decide what's worth doing or not!

Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: `prevActionAsExpected`, `withdrawalCanWithdraw` and `validCommitteTerm` all having `gasAction` as the first argument looks nice and consistent. 

I don't think lambda case is worth it :slightly_smiling_face: 

That being said, it will still change the type of the function, so mentioning it in the changelog is a good practice, but it will not be the end of the world if you don't do it.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs [hunk 2]
```diff
@@ -395,14 +394,17 @@ updatePrevGovActionIdsChildren gas pgac =
...
-  let maxCommitteeTerm = pp ^. ppCommitteeMaxTermLengthL
-      members = foldMap' (^. committeeMembersL) committee
-   in all (<= currentEpoch + maxCommitteeTerm) members
+validCommitteeTerm govAction pp currentEpoch =
+  case govAction of
+    UpdateCommittee _ _ newMembers _ -> valid newMembers
+    _ -> True
+  where
+    maxCommitteeTerm = pp ^. ppCommitteeMaxTermLengthL
+    valid = all (<= currentEpoch + maxCommitteeTerm)
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: `valid` is a little bit too general, maybe it could be a good idea to be more specific with the name?
 
```suggestion
    UpdateCommittee _ _ newMembers _ -> withinMaxTermLength newMembers
    _ -> True
  where
    committeeMaxTermLength = pp ^. ppCommitteeMaxTermLengthL
    withinMaxTermLength = all (<= currentEpoch + committeeMaxTermLength)
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 3]
```diff
@@ -622,10 +625,14 @@ logRatificationChecks gaId = do
     unlines
       [ "----- RATIFICATION CHECKS -----"
       , "prevActionAsExpected:\t" <> show (prevActionAsExpected gasAction ensPrevGovActionIds)
-      , "validCommitteeTerm:\t" <> show (validCommitteeTerm ensCommittee ensCurPParams currentEpoch)
+      , "validCommitteeTerm:\t" <> show (validCommitteeTerm gasAction ensCurPParams currentEpoch)
       , "notDelayed:\t\t??"
       , "withdrawalCanWithdraw:\t" <> show (withdrawalCanWithdraw gasAction curTreasury)
       , "committeeAccepted:\t" <> show (committeeAccepted ratEnv ratSt gas)
+      , "committeeAccepted:\t\t"
+          <> show (committeeAccepted ratEnv ratSt gas)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs]: Duplicate information
```suggestion
      , "committeeAccepted:\t\t"
          <> show (committeeAccepted ratEnv ratSt gas)
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 4]
```diff
@@ -622,10 +625,14 @@ logRatificationChecks gaId = do
...
+          <> " [To Pass, at least: "
+          <> show (votingCommitteeThreshold ratSt gasAction)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:638]: Just like with SPOs and DReps it will be useful to add the actual accepted ratio, not just the threshold:
```suggestion
          <> show (committeeAcceptedRatio ratEnv ratSt gas)
          <> " >= "
          <> show (votingCommitteeThreshold ratSt gasAction)
          <> "]"
```

Note, I am suggesting to change the type signatures (thus implementation) of all three functions to be consistent:

* `committeeAcceptedRatio :: RatifyEnv era -> RatifyState era -> GovActionState era -> Rational`
* `dRepAcceptedRatio :: RatifyEnv era -> GovActionState era -> Rational`
* and `spoAcceptedRatio :: RatifyEnv era -> GovActionState era -> Rational`

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 5]
```diff
@@ -59,15 +62,7 @@ import Cardano.Ledger.BaseTypes (
...
-import Cardano.Ledger.Conway.Core (
-  Era (..),
-  EraIndependentTxBody,
-  EraTx (..),
-  EraTxBody (..),
-  EraTxOut (..),
-  PParams,
-  PParamsHKD,
- )
+import Cardano.Ledger.Conway.Core (DRepVotingThresholds (..), Era (..), EraIndependentTxBody, EraTx (..), EraTxBody (..), EraTxOut (..), PParams, PParamsHKD)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs]: Line is too long. Also it is a Core module, no need to have explicit imports:
```suggestion
import Cardano.Ledger.Conway.Core
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs [hunk 6]
```diff
@@ -272,7 +275,94 @@ spec =
...
+          curConstitution <- getsNES $ newEpochStateGovStateL . constitutionGovStateL
+          curConstitution `shouldBe` constitution
+          pure govIdConst1
+
+        -- after enough epochs pass, the expiration of the new members becomes acceptable
+        -- and the new committee is enacted
+        impAnn "New committee is enacted" $ do
+          currentEpoch <- getsNES nesELL
+          let (EpochNo delta) = exceedingExpiry - maxTermLength - currentEpoch
+          replicateM_ (fromIntegral delta) passEpoch
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs]: EpochNo is a Word64, so this calculation can easily overflow resulting in an "infinite" call in `replicateM_`. Not something you want to run into when debugging tests.
```suggestion
          let delta = fromIntegral exceedingExpiry - fromIntegral (maxTermLength + currentEpoch)
          replicateM_ delta passEpoch
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs [hunk 7]
```diff
@@ -272,7 +275,94 @@ spec =
...
+            let initialMembers = [(c1, maxExpiry), (c2, maxExpiry)]
+            govIdCom1 <-
+              electCommittee
+                SNothing
+                drep
+                Set.empty
+                initialMembers
+            passEpoch >> passEpoch
+            expectMembers [c1, c2]
+            pure ([c1, c2], govIdCom1)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs]: Improve resilience to changes of `initialMembers`
```suggestion
            expectMembers $ Map.keysSet initialMembers
            pure (Map.keysSet initialMembers, govIdCom1)
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs [hunk 8]
```diff
@@ -272,7 +275,94 @@ spec =
...
+            _ <-
+              electCommittee
+                (SJust govIdCom1)
+                drep
+                Set.empty
+                membersExceedingExpiry
+            passEpoch >> passEpoch
+            -- the new committee has not been enacted
+            expectMembers initialMembers
+            pure ([c3, c4], exceedingExpiry)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs]: Ditto
```suggestion
            pure (Map.keysSet initialMembers, exceedingExpiry)
```

**Review by lehins — APPROVED** [review]
Looks great. Although I'd like to suggest few minor improvements.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs (hunk 3 above)
Comment (teodanciu) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs]: Eek! Thanks for catching

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs (hunk 4 above)
Comment (teodanciu) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:638]: Funnily, I'm pretty sure this is why I duplicated the line above, because I intended to achieve exactly this. But then something went wrong... Let me try again. Thanks!

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs (hunk 2 above)
Comment (teodanciu) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: Totally!  Initially it was even worse, I had called it "ok"  :laughing:  But definitely worth investing in good names, thanks for the suggestions.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs (hunk 6 above)
Comment (teodanciu) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs]: Had to apply `unEpochNo` before `fromIntegral`, otherwise I would get  `Could not deduce (Integral EpochNo)
        arising from a use of ‘fromIntegral’`, which make sense i suppose.
